### PR TITLE
mds: add opened files/inodes and pinned i_caps metric payload support

### PR DESCRIPTION
**Pull request by lxbsz**
Fixes: https://tracker.ceph.com/issues/46865
Signed-off-by: Xiubo Li <xiubli@redhat.com>

Depends on: 
https://github.com/ceph/ceph/pull/37654
https://github.com/ceph/ceph/pull/32479

And needs to rebase after them merged.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by lxbsz** [comment]
The depending PRs has been merged and have rebased it to the upstream. Thanks.

**Comment by lxbsz** [comment]
jenkins test api

**Review by batrick — APPROVED** [review]
LGTM, please also run by @vshankar

File: src/client/Client.cc [hunk 1]
```diff
@@ -6415,9 +6422,14 @@ void Client::flush_cap_releases()
       } else {
         session.con->send_message2(std::move(session.release));
       }
+      nr_caps += session.release->caps.size();
```
Comment (batrick) [inline src/client/Client.cc:6425]: `session.release` may be null due to the previous `else` block. It causes this assertion:

````
2020-11-14T02:27:59.720 INFO:teuthology.orchestra.run.smithi164.stderr:ceph_test_trim_caps: /build/ceph-16.0.0-7160-g4090b2dd/obj-x86_64-linux-gnu/boost/include/boost/smart_ptr/intrusive_ptr.hpp:199: T* boost::intrusive_ptr<T>::operator->() const [with T = MClientCapRelease]: Assertion `px != 0' failed.
````

From: /ceph/teuthology-archive/pdonnell-2020-11-14_02:06:06-fs-wip-pdonnell-testing-20201113.235156-distro-basic-smithi/5620640/teuthology.log

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/client/Inode.cc [hunk 2]
```diff
@@ -137,6 +137,9 @@ void Inode::make_nosnap_relative_path(filepath& p)
 
 void Inode::get_open_ref(int mode)
 {
+  client->inc_opened_files();
+  if (open_by_mode.at(mode) == 0)
```
Comment (batrick) [inline src/client/Inode.cc:141]: Also:

````
2020-11-14T02:34:18.568 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr:terminate called after throwing an instance of 'std::out_of_range'
2020-11-14T02:34:18.568 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr:  what():  map::at
2020-11-14T02:34:18.568 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr:*** Caught signal (Aborted) **
2020-11-14T02:34:18.568 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: in thread 7f7e967fc700 thread_name:ceph-fuse
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: ceph version 16.0.0-7160-g4090b2dd (4090b2ddb3668718593e366f72a099e049d8e78a) pacific (dev)
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 1: /lib64/libpthread.so.0(+0x12dc0) [0x7f7ec34b3dc0]
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 2: gsignal()
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 3: abort()
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 4: /lib64/libstdc++.so.6(+0x9006b) [0x7f7ec292806b]
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 5: /lib64/libstdc++.so.6(+0x964cc) [0x7f7ec292e4cc]
2020-11-14T02:34:18.569 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 6: /lib64/libstdc++.so.6(+0x96527) [0x7f7ec292e527]
2020-11-14T02:34:18.570 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 7: /lib64/libstdc++.so.6(+0x96788) [0x7f7ec292e788]
2020-11-14T02:34:18.570 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 8: /lib64/libstdc++.so.6(+0x9200f) [0x7f7ec292a00f]
2020-11-14T02:34:18.570 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 9: (Inode::get_open_ref(int)+0x13c) [0x564e2de72a9c]
2020-11-14T02:34:18.570 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 10: (Client::_create(Inode*, char const*, int, unsigned int, boost::intrusive_ptr<Inode>*, Fh**, int, int, int, char const*, bool*, UserPerm const&)+0x452) [0x564e2de1eae2]
2020-11-14T02:34:18.570 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 11: (Client::_ll_create(Inode*, char const*, unsigned int, int, boost::intrusive_ptr<Inode>*, int, Fh**, UserPerm const&)+0x219) [0x564e2de20089]
2020-11-14T02:34:18.570 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 12: (Client::ll_create(Inode*, char const*, unsigned int, int, stat*, Inode**, Fh**, UserPerm const&)+0xc5) [0x564e2de20a35]
2020-11-14T02:34:18.571 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 13: ceph-fuse(+0x78a36) [0x564e2ddada36]
2020-11-14T02:34:18.571 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 14: /lib64/libfuse.so.2(+0x162ab) [0x7f7ecd5312ab]
2020-11-14T02:34:18.571 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 15: /lib64/libfuse.so.2(+0x17818) [0x7f7ecd532818]
2020-11-14T02:34:18.571 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 16: /lib64/libfuse.so.2(+0x14400) [0x7f7ecd52f400]
2020-11-14T02:34:18.571 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 17: /lib64/libpthread.so.0(+0x82de) [0x7f7ec34a92de]
2020-11-14T02:34:18.571 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi047.stderr: 18: clone()
````

From: /ceph/teuthology-archive/pdonnell-2020-11-14_02:06:06-fs-wip-pdonnell-testing-20201113.235156-distro-basic-smithi/5620648/teuthology.log

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc]: FIxed, thanks.

File: src/client/Inode.cc (hunk 2 above)
Comment (lxbsz) [reply src/client/Inode.cc]: FIxed, thanks.

**Comment by lxbsz** [comment]
@batrick Could you help review this one ? Thanks.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-03-09_03:27:39-fs-wip-pdonnell-testing-20210308.214827-distro-basic-smithi/

Failures are:

- https://tracker.ceph.com/issues/49500
- https://tracker.ceph.com/issues/48805
- https://tracker.ceph.com/issues/48773
- https://tracker.ceph.com/issues/45434
- https://tracker.ceph.com/issues/49240
- https://tracker.ceph.com/issues/49466
- https://tracker.ceph.com/issues/49684
- https://tracker.ceph.com/issues/48771

**Pull request merged**